instance/utils.py

```diff
@@ -73,26 +75,50 @@ def get_requests_retry(total=10, connect=10, read=10, redirect=10, backoff_facto
     )
 
 
-def read_files(*fds):
+def _line_timeout_generator(line_timeout, global_timeout):
     """
-    Given a list of objects implementing the file interface, poll them for new
-    data and yield the lines read as they are written.
+    Helper function for read_files() to compute the timeout for a single line.
+    """
+    if global_timeout is not None:
+        deadline = time.time() + global_timeout
+        while True:
+            global_timeout = deadline - time.time()
+            if line_timeout is not None:
+                yield min(line_timeout, global_timeout)
+            else:
+                yield global_timeout
+    else:
+        yield from itertools.repeat(line_timeout)
+
+
+def read_files(*files, line_timeout=None, global_timeout=None):
```

Comment:
Maybe we should rename this function.  `read_files()` doesn't seem a good name, given that the main thing it does is multiplexing and not reading, and it doesn't even work for files. :)